fastly/data_source_fastly_tls_certificate.go

```diff
@@ -0,0 +1,223 @@
+				Optional:      true,
+				Computed:      true,
+				ConflictsWith: []string{"id"},
+			},
+			"issued_to": {
+				Type:          schema.TypeString,
+				Description:   "The hostname for which a certificate was issued.",
+				Optional:      true,
+				Computed:      true,
+				ConflictsWith: []string{"id"},
+			},
+			"domains": {
+				Type:          schema.TypeSet,
+				Description:   "Domains that are listed in any certificates' Subject Alternative Names (SAN) list.",
+				Optional:      true,
+				Computed:      true,
+				Elem:          &schema.Schema{Type: schema.TypeString},
+				ConflictsWith: []string{"id"},
+			},
+			"issuer": {
+				Type:          schema.TypeString,
+				Description:   "The certificate authority that issued the certificate.",
+				Optional:      true,
+				Computed:      true,
+				ConflictsWith: []string{"id"},
+			},
+			"created_at": {
+				Type:        schema.TypeString,
+				Description: "Timestamp (GMT) when the certificate was created",
+				Computed:    true,
+			},
+			"updated_at": {
+				Type:        schema.TypeString,
+				Description: "Timestamp (GMT) when the certificate was last updated",
+				Computed:    true,
+			},
+			"replace": {
+				Type:        schema.TypeBool,
+				Description: "A recommendation from Fastly indicating the key associated with this certificate is in need of rotation",
+				Computed:    true,
+			},
+			"serial_number": {
+				Type:        schema.TypeString,
+				Description: "A value assigned by the issuer that is unique to a certificate",
+				Computed:    true,
+			},
+			"signature_algorithm": {
+				Type:        schema.TypeString,
+				Description: "The algorithm used to sign the certificate",
+				Computed:    true,
+			},
+		},
+	}
+}
+
+func dataSourceFastlyTLSCertificateRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*FastlyClient).conn
+
+	var certificate *fastly.CustomTLSCertificate
+
+	if v, ok := d.GetOk("id"); ok {
+		cert, err := conn.GetCustomTLSCertificate(&fastly.GetCustomTLSCertificateInput{
+			ID: v.(string),
+		})
+		if err != nil {
+			return err
+		}
+
+		certificate = cert
+	} else {
+		filters := getTLSCertificateFilters(d)
+
+		certificates, err := listTLSCertificates(conn, filters...)
+		if err != nil {
+			return err
+		}
+
+		if len(certificates) == 0 {
+			return fmt.Errorf("Your query returned no results. Please change your search criteria and try again.")
+		}
+
+		if len(certificates) > 1 {
+			return fmt.Errorf("Your query returned more than one result. Please change try a more specific search criteria and try again.")
+		}
+
+		certificate = certificates[0]
+	}
+
+	return dataSourceFastlyTLSCertificateSetAttributes(certificate, d)
+}
+
+type TLSCertificatePredicate func(*fastly.CustomTLSCertificate) bool
+
+func getTLSCertificateFilters(d *schema.ResourceData) []TLSCertificatePredicate {
```

Comment:
This was definitely one that seemed a good candidate for generics, agreed! I don't believe so, no, if we maintain the "_reasonable_" caveat. Off the top of my head the main option would be to pass around `interface{}`s instead of the `fastly.*` types and either forfeit type safety or have a load of type switches around obscuring the logic. If you think of anything, let me know, otherwise I guess we're waiting until December or whenever generics gets added!

Reply:
👍🏻